# 🌳 Comparación de Modelos: ForestGroot YOLOv8 vs. YOLOv9

## Integrantes:
- **Juan Camilo Vargas**
- **Samuel Pinzón Valderrutén**
- **Sebastian Diaz Noguera**

![ForestGroot](https://i.ibb.co/WzNPDXN/Sin-t-tulo.png)

## **Dataset Forestgroot**

Para la creación de este dataset se recolectaron un total de 400 imágenes JPG de zonas adyacentes al Amazonas utilizando imágenes de [Google Earth](https://www.google.com/earth/). A continuación, se presenta una tabla que describe las clases identificadas en las imágenes recolectadas:

<div align="center">

| **Clase**       | **Descripción**                                        |
|-----------------|--------------------------------------------------------|
| deforestation   | Áreas donde se observa la pérdida de bosques.          |
| water           | Regiones que contienen ríos y cuerpos de agua.         |
| fire            | Zonas afectadas por incendios forestales.              |
| urban           | Áreas urbanas, incluyendo ciudades y asentamientos urbanos. |

</div>

Debido al desbalance de clase con las 400 imágenes se realizó un data augmentation en las siguientes clases para evitar este problema:

- **urban:** 245 imágenes.
- **fire:** 85 imágenes.
- **water:** 100 imágenes.

Se obtuvo un dataset con un total de **830** imágenes, donde se identificaron las siguientes cantidades después del etiquetado:

<div align="center">

| Clase          | Cantidad |
|----------------|----------|
| deforestation  | 558      |
| urban          | 511      |
| fire           | 477      |
| water          | 470      |

</div>

Posteriormente, se realizó el data augmentation con Roboflow y se obtuvo el dataset final con **1992** imágenes, distribuidas de la siguiente manera:

- Conjunto de entrenamiento: **1743** imágenes.
- Conjunto de prueba: **166** imágenes.
- Conjunto de validación: **83** imágenes.

**Ejemplo del Dataset:**

<p align="center">
  <img src="https://i.ibb.co/nrfJFSW/deforestation-098.png" alt="ForestGroot" width="70%">
</p>


##**YOLOv8**
YOLOv8 es una versión de la familia de algoritmos de detección de objetos conocida como You Only Look Once (YOLO). Este modelo destaca por su capacidad para equilibrar velocidad y precisión en tareas de visión artificial en tiempo real. YOLOv8 sigue este enfoque al no contar con capas densas, todas sus capas son convolucionales. Esto lo hace adecuado para escenarios donde el procesamiento en tiempo real es crucial.

**Proceso de detección de objetos en YOLOv8:**

1. **Extracción de características:**
   - La imagen se divide en una cuadrícula de celdas, y luego se utiliza una red neuronal convolucional para extraer características de cada una.
   - Estas características representan la información visual dentro de cada celda.

2. **Predicción de Cajas Delimitadoras y Clases:**
   - Para cada celda de la cuadrícula, YOLOv8 predice un número fijo de cajas delimitadoras y las probabilidades de las clases correspondientes.
   - Cada caja delimitadora contiene información sobre la posición (coordenadas x, y, ancho y alto) y la confianza de detección.

3. **Supresión de No Máximos (NMS):**
   - Se aplica NMS para eliminar detecciones redundantes y mantener solo las detecciones más confiables.
   - Durante este proceso, se selecciona la caja delimitadora con la mayor probabilidad de objeto y se eliminan las demás cajas que tienen una superposición significativa con ella.

4. **Salida de Detecciones:**
   - La salida final consiste en las cajas delimitadoras seleccionadas después de la supresión de no máximos, junto con las clases detectadas y sus respectivas probabilidades.

**Arquitectura:**

![yolov8_arciteture](https://user-images.githubusercontent.com/62583018/211719362-39fc8a88-b1ce-4ab3-9a9f-b640550515b4.jpg)

## Roboflow

**[Roboflow](https://roboflow.com/)** es una plataforma que facilita el proceso de preparación de datos para proyectos de visión por computadora y aprendizaje automático. Proporciona herramientas para organizar, limpiar, etiquetar y aumentar conjuntos de datos de imágenes de forma eficiente. Además, facilita la creación de modelos de aprendizaje automático a través de su plataforma.

La preparación de un conjunto de datos personalizado puede representar un desafío considerable, pero como se mencionó, Roboflow simplifica enormemente este proceso. A continuación, se describen los pasos seguidos para llevar a cabo este procedimiento:

1. **Creación del Proyecto:** Se inició un proyecto específico para la detección de objetos en Roboflow.
2. **Subida de Imágenes:** Se subieron las imágenes recolectadas previamente al proyecto creado.
3. **Etiquetado de Clases:** Se realizó el etiquetado de las clases de interés en las imágenes, como deforestación, agua, fuego y urbano.
4. **Aumento de Datos:** Se aplicaron técnicas de aumento de datos para mejorar la diversidad y cantidad de imágenes disponibles.
5. **Exportación del Dataset:** Finalmente, se exportó el dataset preparado en el formato adecuado para su uso en el entorno de entrenamiento.

[![Roboflow](https://i.ibb.co/kMrgGYL/Sin-t-tulo.png)](https://ibb.co/CWX5wxj)



## **Detección de Deforestación**

**Descarga de las Librerías**

In [ ]:
# Instalación de paquetes necesarios
!pip install ultralytics -q
!pip install roboflow -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 779.6/779.6 kB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.3/21.3 MB 54.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 75.5/75.5 kB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.3/158.3 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 178.7/178.7 kB 8.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 6.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.1/49.1 MB 11.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 6.8 MB/s eta 0:00:00


In [ ]:
# Importación de librerías
import glob
import os

from IPython.display import display, Image
from roboflow import Roboflow
import tensorflow as tf
from ultralytics import YOLO

**Entorno de Trabajo**

Es importante asegurarse de contar con acceso a una unidad de procesamiento gráfico (GPU) para garantizar un rendimiento óptimo durante el proceso.

In [ ]:
# Verificar si se encuentra disponible la GPU
device_name = tf.test.gpu_device_name()
# Si no se encuentra disponible la GPU, lanzar una excepción
if device_name != '/device:GPU:0':
    raise SystemError('No se encontró dispositivo GPU')
# Imprimir el nombre del dispositivo GPU encontrado
print('Dispositivo GPU encontrado en: {}'.format(device_name))

Dispositivo GPU encontrado en: /device:GPU:0


In [ ]:
# Define el directorio de trabajo actual
HOME = os.getcwd()
HOME

'/content'

**Dataset desde Roboflow**

Como se mencionó anteriormente, se utilizó la herramienta de Roboflow para la creación del dataset. Para acceder al dataset, se puede utilizar la API de la siguiente manera. En caso de que no puedas acceder a la API, también puedes descomprimir el archivo con el dataset que se encuentra en el repositorio.

In [ ]:
!mkdir {HOME}/datasets
%cd {HOME}/datasets

# Descargar el conjunto de datos de Roboflow utilizando la API de Roboflow
rf = Roboflow(api_key="w2EZPylLkOYJSnI4o6qN")
project = rf.workspace("test-yyciu").project("groot-v2")
version = project.version(3)
dataset = version.download("yolov8")

/content/datasets
loading Roboflow workspace...
loading Roboflow project...
Dependency ultralytics==8.0.196 is required but found version=8.2.28, to fix: `pip install ultralytics==8.0.196`



Extracting Dataset Version Zip to Groot-V2-3 in yolov8:: 100%|██████████| 3996/3996 [00:00<00:00, 4300.30it/s]


**Validación**

In [ ]:
%cd {HOME}
# Validar los modelos
models_path = f'{HOME}/models'

for forestgroot_model in os.listdir(models_path):
    if not forestgroot_model.lower().endswith('.pt'):
      continue
    model_name = ' '.join(forestgroot_model.split('_')[:2])
    print(f"****** Model: {model_name} ******")
    model_path = os.path.join(models_path, forestgroot_model)
    forestgroot_seg = YOLO(model_path)
    validation_results = forestgroot_seg.val()
    print("****************************************\n")

/content
****** Model: forestgroot yolov8s ******
Ultralytics YOLOv8.2.28 🚀 Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
YOLOv8s-seg summary (fused): 195 layers, 11781148 parameters, 0 gradients, 42.4 GFLOPs


100%|██████████| 755k/755k [00:00<00:00, 23.8MB/s]
val: Scanning /content/datasets/Groot-V2-3/valid/labels... 166 images, 1 backgrounds, 0 corrupt: 100%|██████████| 166/166 [00:00<00:00, 1343.33it/s]

val: New cache created: /content/datasets/Groot-V2-3/valid/labels.cache



os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:12<00:00,  1.09s/it]


                   all        166        415       0.75      0.718      0.769      0.611       0.73      0.731      0.765      0.573
         deforestation         37        106      0.826      0.623      0.763      0.679      0.784      0.632       0.75      0.638
                  fire         39        100      0.695      0.773      0.789      0.665       0.69        0.8      0.801      0.632
                 urban         68         93      0.703       0.71      0.758      0.474      0.685       0.72      0.741      0.417
                 water         75        116      0.775      0.767      0.767      0.627      0.762      0.773      0.768      0.604
Speed: 9.1ms preprocess, 13.1ms inference, 0.0ms loss, 9.6ms postprocess per image
Results saved to runs/segment/val2
****************************************

****** Model: forestgroot yolov9c ******
WARNING ⚠️ /content/models/forestgroot_yolov9c_seg.pt appears to require 'dill', which is not in ultralytics requirements.
AutoInstall

val: Scanning /content/datasets/Groot-V2-3/valid/labels.cache... 166 images, 1 backgrounds, 0 corrupt: 100%|██████████| 166/166 [00:00<?, ?it/s]
os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:11<00:00,  1.07s/it]


                   all        166        415      0.768      0.755      0.782      0.639      0.772      0.754      0.785       0.58
         deforestation         37        106      0.762      0.689       0.75      0.676      0.758      0.679      0.733      0.619
                  fire         39        100      0.717       0.77      0.796      0.675      0.732       0.78      0.808      0.626
                 urban         68         93      0.747      0.763      0.776      0.515      0.761      0.774      0.797      0.448
                 water         75        116      0.845      0.798      0.805       0.69      0.839      0.784      0.805      0.626
Speed: 3.2ms preprocess, 36.6ms inference, 0.0ms loss, 3.4ms postprocess per image
Results saved to runs/segment/val3
****************************************

****** Model: forestgroot yolov8n ******
Ultralytics YOLOv8.2.28 🚀 Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
YOLOv8n-seg summary (fused): 195 layers, 325884

val: Scanning /content/datasets/Groot-V2-3/valid/labels.cache... 166 images, 1 backgrounds, 0 corrupt: 100%|██████████| 166/166 [00:00<?, ?it/s]
os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:11<00:00,  1.01s/it]


                   all        166        415      0.785      0.703      0.771      0.609      0.789      0.708      0.775      0.573
         deforestation         37        106      0.766      0.698      0.742      0.659      0.754      0.689       0.74      0.616
                  fire         39        100      0.805        0.7        0.8      0.675      0.816      0.712      0.807      0.645
                 urban         68         93      0.747      0.688      0.762      0.492      0.756      0.701      0.771      0.449
                 water         75        116      0.823      0.724       0.78      0.611      0.831      0.733      0.782      0.579
Speed: 4.6ms preprocess, 10.7ms inference, 0.0ms loss, 5.8ms postprocess per image
Results saved to runs/segment/val4
****************************************

****** Model: forestgroot yolov8m ******
Ultralytics YOLOv8.2.28 🚀 Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
YOLOv8m-seg summary (fused): 245 layers, 272247

val: Scanning /content/datasets/Groot-V2-3/valid/labels.cache... 166 images, 1 backgrounds, 0 corrupt: 100%|██████████| 166/166 [00:00<?, ?it/s]
os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:11<00:00,  1.01s/it]


                   all        166        415      0.724      0.736      0.773      0.627      0.735      0.747      0.775      0.576
         deforestation         37        106      0.736      0.689      0.751      0.689      0.737      0.689      0.749      0.625
                  fire         39        100       0.66      0.737      0.774      0.631      0.659      0.735      0.773       0.61
                 urban         68         93      0.717      0.753      0.759      0.494      0.758      0.796      0.776       0.45
                 water         75        116      0.783      0.767      0.806      0.693      0.783      0.767      0.803      0.619
Speed: 4.5ms preprocess, 28.2ms inference, 0.2ms loss, 3.7ms postprocess per image
Results saved to runs/segment/val5
****************************************

****** Model: forestgroot yolov8l ******
Ultralytics YOLOv8.2.28 🚀 Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
YOLOv8l-seg summary (fused): 295 layers, 459149

val: Scanning /content/datasets/Groot-V2-3/valid/labels.cache... 166 images, 1 backgrounds, 0 corrupt: 100%|██████████| 166/166 [00:00<?, ?it/s]
os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:12<00:00,  1.17s/it]


                   all        166        415      0.786      0.729      0.784      0.642      0.782      0.725       0.78      0.585
         deforestation         37        106      0.803      0.693      0.765      0.699      0.803      0.693      0.766      0.645
                  fire         39        100      0.745      0.758      0.805       0.67      0.745      0.759      0.804      0.643
                 urban         68         93      0.766      0.656      0.758      0.504      0.765      0.656      0.761      0.428
                 water         75        116      0.831       0.81      0.807      0.694      0.813      0.793      0.788      0.625
Speed: 3.3ms preprocess, 43.4ms inference, 0.0ms loss, 3.4ms postprocess per image
Results saved to runs/segment/val6
****************************************



**Resultados del Modelo ForestGroot:**
<div align="center">

| Modelo               | mAP50 <sub>box</sub> deforestation | mAP50<sub>mask</sub> deforestation | Params (M) |
|----------------------|:----------------------------------:|:---------------------------------:|:----------:|
| ForestGroot Yolov8n  |                0.742               |               0.74                |    3.26    |
| ForestGroot Yolov8s  |               **0.763**            |               0.75                |   11.78    |
| ForestGroot Yolov8m  |                0.751               |               0.749               |   27.22    |
| ForestGroot Yolov8l  |               **0.765**            |             **0.766**             |   45.91    |
| ForestGroot Yolov9c  |                0.75                |               0.73                |   27.63    |

</div>


## Elección del Modelo Final

Para la elección del modelo final, después de realizar la validación de los modelos, tuvimos en cuenta principalmente la métrica **mAP50** (mean Average Precision at IoU=0.50). Los modelos que mejor rendimiento obtuvieron fueron el **ForestGroot YOLOv8L** y el **ForestGroot YOLOv9c**, ambos con un rendimiento de **0.78** en todas las clases, tanto en la detección de boxes como en máscaras (masks).

Sin embargo, dado que nuestro principal objetivo es la detección de la clase **deforestation**, la decisión final se basó en el rendimiento específico en esta clase. En este aspecto, el modelo **ForestGroot YOLOv8L** tuvo un rendimiento superior, alcanzando un **0.76** tanto en la detección de boxes como en máscaras. En comparación, el **ForestGroot YOLOv9c** obtuvo un **0.75** en boxes y un **0.73** en máscaras para la clase **deforestation**.

Debido a esta diferencia en la clase más importante para nuestro proyecto, decidimos que el modelo **ForestGroot YOLOv8L** es el más adecuado para nuestro objetivo.
